[Enhancement] Speed up Gin writing

srlch opened the pull request:
This pr introduce some optimization for speeding up
Gin writing:
1. Use ref value pass for the wchat_t data.
2. Omit to write the norm information for the index.
3. Turn ON the AVX2 option for gin writing for the faster string comparation.



Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

wyb commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -121,8 +122,25 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
+                const char* s = _val->data;
+                size_t size = _val->size;
+                if (size > _wstring_buffer.size()) {
+                    _wstring_buffer.resize(size);
```
`reserve`?

srlch replied:
`reserve `may make the size() unchanged. We can not get the current _wstring_buffer size in this case. But we need to change it in every loop.

srlch replied:
use reserve instead

wyb commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -121,8 +122,25 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
+                    _field->setValue(stream);
+                    _index_writer->addDocument(_doc.get());
+                } else {
+                    _field->setValueRef(data);
```
`_wstring_buffer` will be rewrited in the next loop, is `data ref` right?

kevincai commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -224,6 +230,7 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
+    std::wstring _wstring_buffer;
 };
 
 Status CLuceneInvertedWriter::create(const TypeInfoPtr& typeinfo, const std::string& field_name,
```
this is a valid review comment.

srlch replied:
Use raw byte buffer instead.

kevincai commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -224,6 +230,7 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
     InvertedIndexParserType _parser_type;
     std::wstring _field_name;
     const TabletIndex* _inverted_index;
+    std::wstring _wstring_buffer;
```
if wstring_buffer is just used for a continuous buffer, shall use raw byte or raw char array instead of wstring, to avoid the overhead of initializing the string.

srlch replied:
Use raw byte buffer instead.

kevincai commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -45,7 +45,7 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
     explicit CLuceneInvertedWriterImpl(const std::string& field_name, const std::string& directory,
                                        const TabletIndex* inverted_index)
-            : _directory(std::move(directory)), _inverted_index(inverted_index) {
+            : _directory(std::move(directory)), _inverted_index(inverted_index), _wstring_buffer(1024 * 1024, '\0') {
```
`_wstring_buffer(1024 * 1024, '\0')` this will introduce a memset of 2MB memory for every _wstring_buffer.

kevincai commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -121,8 +122,25 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
+                    _wstring_buffer.reserve(size);
+                }
+                TCHAR* data = const_cast<TCHAR*>(_wstring_buffer.data());
+                std::mbstowcs(data, s, size);
```
this will write data after its _wstring.size(), even though the underlying buffer has enough space. It is not suggested to do it in this hacked way that completely breaks the std::string semantics.

srlch replied:
same as above

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -1210,6 +1210,16 @@ build_libdeflate() {
 
 #clucene
 build_clucene() {
+    if [[ -z ${USE_AVX2} ]]; then
```
check if clucene can be enhanced to build avx2 and non-avx2 version co-exist in the library, and determine which one to use during the runtime where the cpuset info can be leveraged.

srlch replied:
use a common flag for the whole script

kevincai commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -224,6 +235,8 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
     InvertedIndexParserType _parser_type;
     std::wstring _field_name;
     const TabletIndex* _inverted_index;
+    void* _wstring_buffer;
```
shall guarantee the buffer will be freed in case of any errors or early exit.

srlch replied:
Fixed it

kevincai commented on be/src/storage/inverted/clucene/clucene_inverted_writer.cpp:
```diff
@@ -49,6 +49,8 @@ class CLuceneInvertedWriterImpl : public CLuceneInvertedWriter {
         _parser_type = get_inverted_index_parser_type_from_string(
                 get_parser_string_from_properties(_inverted_index->index_properties()));
         _field_name = std::wstring(field_name.begin(), field_name.end());
+        _wstring_buffer = (void*)malloc(2 * 1024 * 1024);
```
can this buffer can be allocated as near as its usage? so the mem is not necessary whenever a CLuceneInvertedWriterImpl object is created.

srlch replied:
This is done intentionally to allocate more memory in advance. The purpose here is to ensure that the buffer is large enough to accommodate the contents in most rows and avoid frequent memory reallocation.

kevincai replied:
that's fine to allocate memory all at once large enough. But it is not necessary to allocate when constructing the object.

E.g. when do things like `auto a = std::vector<CLuceneInvertedWriterImpl>(10)` and then `a.clear()`, these mem allocation is useless. allocating the memory as near as to its usage, make the constructing of the object as lightweight as possible.

srlch replied:
Fixed it, allocate memory when we use it.

srlch replied:
refactor the code for correctness.

srlch committed:
[Enhancement] Speed up Gin writing

This pr introduce some optimization for speeding up
Gin writing:
1. Use ref value pass for the wchat_t data.
2. Omit to write the norm information for the index.
3. Turn ON the AVX2 option for gin writing for the
   faster string comparation.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix

Signed-off-by: srlch <linzichao@starrocks.com>
